File: packages/sdk/js/packages/admin/src/d1.ts
/**
 * D1Client — SDK client for user-defined D1 databases
 *
 * Usage:
 *   const admin = createAdminClient(url, { serviceKey });
 *   const rows = await admin.d1('analytics').exec('SELECT * FROM events WHERE type = ?', ['pageview']);
 */
import type { HttpClient } from '@edgebase/core';
import { HttpClientAdapter } from '@edgebase/core';
import { DefaultAdminApi } from './generated/admin-api-core.js';

export class D1Client {
  private adminCore: DefaultAdminApi;
  private database: string;

  constructor(httpClient: HttpClient, database: string) {
    this.adminCore = new DefaultAdminApi(new HttpClientAdapter(httpClient));
    this.database = database;
  }

  /**
   * Execute a SQL query on the D1 database.
   * All SQL is allowed (DDL included) — Service Key holders are admin-level trusted.
   * Use ? placeholders for bind parameters (SQL injection prevention).
   *
   * @param query SQL query string with ? placeholders
   * @param params Bind parameters
   * @returns Array of result rows
   */
  async exec<T = Record<string, unknown>>(query: string, params?: unknown[]): Promise<T[]> {
    const res = await this.adminCore.executeD1Query(this.database, {
      query,
      params,
    }) as { results: T[] };
    return res.results;
  }

  /**
   * Alias for `exec()` — execute a SQL query on the D1 database.
   */
  query<T = Record<string, unknown>>(sql: string, params?: unknown[]): Promise<T[]> {
    return this.exec<T>(sql, params);
  }
}

File: packages/sdk/rust/packages/admin/src/d1.rs
//! D1Client — D1 database access for server-side use.

use std::collections::HashMap;
use std::sync::Arc;

use serde_json::Value;

use crate::generated::admin_api_core::GeneratedAdminApi;
use edgebase_core::error::Error;
use edgebase_core::http_client::HttpClient;

/// Client for a user-defined D1 database.
pub struct D1Client {
    http: Arc<HttpClient>,
    pub(crate) database: String,
}

impl D1Client {
    pub(crate) fn new(http: Arc<HttpClient>, database: &str) -> Self {
        Self {
            http,
            database: database.to_string(),
        }
    }

    fn core(&self) -> GeneratedAdminApi<'_> {
        GeneratedAdminApi::new(&self.http)
    }

    /// Execute a SQL query. Use ? placeholders for bind parameters.
    /// All SQL is allowed (DDL included).
    pub async fn exec(&self, query: &str, params: &[Value]) -> Result<Vec<HashMap<String, Value>>, Error> {
        let body = serde_json::json!({
            "query": query,
            "params": params,
        });
        let res: serde_json::Value = self.core().execute_d1_query(&self.database, &body).await?;
        if let Some(results) = res.get("results").and_then(|v: &serde_json::Value| v.as_array()) {
            Ok(results
                .iter()
                .filter_map(|v: &serde_json::Value| {
                    v.as_object()
                        .map(|o: &serde_json::Map<String, serde_json::Value>| o.iter().map(|(k, v): (&String, &serde_json::Value)| (k.clone(), v.clone())).collect::<HashMap<String, serde_json::Value>>())
                })
                .collect())
        } else {
            Ok(vec![])
        }
    }

    /// Alias for exec() to match SDK parity across runtimes.
    pub async fn query(&self, query: &str, params: &[Value]) -> Result<Vec<HashMap<String, Value>>, Error> {
        self.exec(query, params).await
    }
}
